[Hooks] Fix token refresh - @W-12704554@

vmarta opened the pull request:
We have `pendingToken` in our auth logic that was meant to be temporary. But we did not remove it when no longer pending.

Todos:
- [x] write a new test
- [x] clean up code

How to test:
1. Load http://localhost:3000/
2. Remove the access token from your local storage.. for example`RefArchGlobal_access_token`.
3. Navigate to another page
4. Verify in the browser's Network tab that there is a request to the token endpoint, and all the other SCAPI requests are successful.

# Types of Changes

- [x] **Bug fix** (non-breaking change that fixes an issue)
- [ ] **New feature** (non-breaking change that adds functionality)
- [ ] **Documentation update**
- [ ] **Breaking change** (could cause existing functionality to not work as expected)
- [ ] **Other changes** (non-breaking changes that does not fit any of the above)

> Breaking changes include:
>
> - Removing a public function or component or prop
> - Adding a required argument to a function
> - Changing the data type of a function parameter or return value
> - Adding a new peer dependency to `package.json`

# Changes

- (change1)

# How to Test-Drive This PR

- (step1)

# Checklists

## General

- [ ] Changes are covered by test cases
- [ ] CHANGELOG.md updated with a short description of changes (_not_ required for documentation updates)

## Accessibility Compliance

You must check off all items in **one** of the follow two lists:

- [ ] There are no changes to UI

_or..._

- [ ] Changes were tested with a Screen Reader (iOS VoiceOver or Android Talkback) and had no issues
- [ ] Changes comply with [WCAG 2.0 guidelines levels A and AA](https://www.wuhcag.com/wcag-checklist/)
- [ ] Changes to common UI patterns and interactions comply with [WAI-ARIA best practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

## Localization

- [ ] Changes include a UI text update in the Retail React App (which requires translation)

vmarta committed:
Draft a fix to make sure token refresh works as expected

We have pendingToken that was meant to be temporary. But we did not remove it when no longer pending.

vmarta committed:
Einstein events now wait for token to refresh if expired

kevinxh commented on packages/commerce-sdk-react/src/hooks/useAccessToken.ts:
```diff
@@ -23,4 +23,10 @@ const useAccessToken = (): string | null => {
     return auth.get('access_token')
 }
 
+export const useAccessTokenAsync = async () => {
+    const auth = useAuthContext()
+    const {access_token} = await auth.ready()
+    return access_token
+}
```
I have never seen async React hook, is it actually working? it would block the rendering of the components, right?

kevinxh replied:
I'm just thinking about `useAccessToken` v.s. `useAccessTokenAsync`.. I feel like as an end user, I always want to use `useAccessTokenAsync`, there is no reason to have `useAccessToken`

kevinxh replied:
what if we change the `useAccessToken` to return the `ready` promise by default, and remove `useAccessTokenAsync`

kevinxh replied:
Actually, you might need access to both sync and async token... 

dumb ideas...

```suggestion
export const useAccessTokenAsync = async () => {
    const auth = useAuthContext()
    return {token: auth.get('access_token'), ready: auth.ready}
}
```

This way, you have access to both the token declaratively, and also the promise in case if you need to use it in action handlers.

kevinxh commented on packages/template-retail-react-app/app/hooks/use-einstein.js:
```diff
@@ -389,7 +394,7 @@ export class EinsteinAPI {
 const useEinstein = () => {
     const api = useCommerceApi()
-    const token = useAccessToken()
+    const tokenPromise = useAccessTokenAsync()
```
👍 👍 👍 good catch for the async use case

kevinxh commented on packages/commerce-sdk-react/src/auth/index.ts:
```diff
@@ -293,27 +297,31 @@ class Auth {
-            return this.pendingToken
+            return Promise.resolve(this.data)
         }
+        // TODO: in my scenario, the code exits early here. It should continue on and refresh the token.
```
are we lacking tests or is the tests not written properly to catch this? either way, please add test for this bug, thank you!!! 🙇

vmarta replied:
That's a good point. I'm not going to make a _new_ test because the existing ones are comprehensive already. 

It's just that in the testing of expired token, the bug was not caught because the test did not simulate `pendingToken` being set when initially the token was valid.

I have updated the existing tests to verify:
- `pendingToken` is no longer being set from multiple places.. only `queueRequest()` is supposed to set and delete it.
- the refresh-token logic by simulating real-world scenario (initially having a valid token, and then with expired token)

vmarta committed:
Refactor useAccessToken hook

Now only 1 hook and not 2.

vmarta committed:
Delete duplicate test

vmarta committed:
Update existing tests to verify the refresh-token logic

vmarta commented on packages/commerce-sdk-react/src/auth/index.test.ts:
```diff
@@ -133,17 +133,6 @@ describe('Auth', () => {
         expect(newAuth.get('access_token')).not.toBe('123')
         expect(newAuth.get('refresh_token_guest')).not.toBe('456')
     })
-    test('isTokenExpired', () => {
```
Deleting a duplicate test

vmarta commented on packages/commerce-sdk-react/src/auth/index.ts:
```diff
@@ -293,27 +297,31 @@ class Auth {
-            this.pendingToken = Promise.resolve(this.data)
-            return this.pendingToken
+            console.log('--- has access token that is not expired')
+            return Promise.resolve(this.data)
```
No longer setting `pendingToken` here because it's meant to be set when making an API request. Not when the token is already available.

vmarta commented on packages/commerce-sdk-react/src/hooks/useAccessToken.ts:
```diff
@@ -10,17 +10,26 @@ import useConfig from './useConfig'
+
+    const tokenWhenReady = auth.ready().then(({access_token}) => access_token)
+
+    return {token, tokenWhenReady}
```
With Kevin's feedback, now there's only 1 hook for this `useAccessToken`.

vmarta committed:
Bump up the main.js bundle size a bit

vmarta committed:
Merge branch 'feature/integrate-commerce-sdk-react' into fix-token-refresh

vmarta committed:
Increase jest timeout accordingly

To be consistent with what's in the other test file "wishlist-primary-action.test.js".

vmarta committed:
Remove logging

vmarta committed:
Fix tests from running into infinite re-renders

vmarta committed:
Add comment

vmarta committed:
Apply this Einstein-event mock globally

wjhsf commented on packages/commerce-sdk-react/src/auth/index.ts:
```diff
@@ -298,17 +302,15 @@ class Auth {
             this.set('customer_type', isGuest ? 'guest' : 'registered')
-            this.pendingToken = Promise.resolve(this.data)
-            return this.pendingToken
+            return Promise.resolve(this.data)
```
```suggestion
            return this.data
```
The return value from an async function is automatically wrapped in `Promise.resolve`, so you don't have to do it yourself.

wjhsf commented on packages/commerce-sdk-react/src/auth/index.ts:
```diff
@@ -298,17 +302,15 @@ class Auth {
         if (accessToken && !this.isTokenExpired(accessToken)) {
-            this.pendingToken = Promise.resolve(this.data)
-            return this.pendingToken
+            return Promise.resolve(this.data)
```
```suggestion
            return this.data
```
The return value from an async function is automatically wrapped in `Promise.resolve`, so you don't have to do it yourself.

wjhsf commented on packages/template-retail-react-app/app/pages/product-detail/index.test.js:
```diff
@@ -123,7 +120,7 @@ describe('product set', () => {
             {timeout: 5000}
         )
-    })
+    }, 30000)
```
That seems like a _big_ jump for the timeout! Is there any better way to address it?

vmarta replied:
That's big timeout, yes.. but it's not the first time we've done it in the project ;)

On the other hand, is there a better way? Well, Kevin mentioned this last time: investigate why modal is rendering slowly during testing, especially in CI environment. It's something that we haven't had a chance to do, but we can revisit it for sure.

wjhsf commented on packages/commerce-sdk-react/src/hooks/useAccessToken.ts:
```diff
@@ -10,17 +10,28 @@ import useConfig from './useConfig'
+        : auth.get('access_token')
+
+    // NOTE: auth.ready() is to be called later. If you call it immediately in this hook,
+    // it'll cause infinite re-renders during testing.
```
Does it cause infinite re-renders in production as well? If it does, this is fine. If it doesn't, it feels like we should change the tests, rather than this.

vmarta replied:
If "production" means loading retail-react-app in my browser, then I did not see the infinite re-renders there. 

However, I'm not sure if things would really work as well. After all, I was only navigating to just a few pages in the browser. The change I'm making to `useAccessToken` would affect all pages that call the `useEinstein` hook, which means the impact is global. So in this case, I prefer to rely on the tests to tell me if something goes wrong or not.

Additionally, from what I've seen how `auth.ready()` is being called in the project, the biggest takeaway for me is this: `auth.ready()` is meant to be called later, like in an event handler. So our hook needs to first wrap it inside another function.

kevinxh replied:
Today `auth.ready()` is being called from 2 places:

1. on CommerceAPIProvider mount, in a `useEffect` (this is really optional, just a nit performance thing)
2. all SCAPI requests are gated by auth.ready()

kevinxh replied:
> the biggest takeaway for me is this: auth.ready() is meant to be called later

🤔 Where/when you call `auth.ready()` in user land should not be concern! If there is a limitation, we should fix it.

vmarta committed:
Remove unnecessary Promise.resolve

Co-authored-by: Will Harney <62956339+wjhsf@users.noreply.github.com>
